### PR TITLE
Add support for PodDisruptionBudget 

### PR DESCRIPTION
**Pull request by msolimans**
Add support to PodDisruptionBudget to increase HA of ExternalSecrets deployments.

What is the added value?
Achieve HA for ExternalSecrets deployments

Give us examples of the outcome

Add pdb to certController
Add pdb to webhook
Add pdb to operator

A fix to the issue reported in https://github.com/external-secrets/external-secrets/issues/1162

### CODE REVIEW

**Comment by moolen** [comment]
Hey, thanks for bringing this up. This looks good, could you please check that all files are generated correctly? (`make reviewable`)

**Comment by msolimans** [comment]
@moolen Thanks for taking the time to review - I just fixed trailing spaces reported from linting

**Comment by msolimans** [comment]
note:  I couldn't run `make reviewable` in mac as it gave me a couple of unrelated go errors.

**Comment by gusfcarvalho** [comment]
Hi @msolimans ! Can you share the errors that were happening there? We are trying to keep the experience for both linux and mac contributors equally good =)

**Comment by msolimans** [comment]
> Hi @msolimans ! Can you share the errors that were happening there? We are trying to keep the experience for both linux and mac contributors equally good =)

let me run that again and post it here

**Comment by msolimans** [comment]
@gusfcarvalho  I saw a fix to gnu-sed was recently merged (today in the morning) and that solved the below error: 

```
~/VisualStudioProjects/external-secrets main ⇡7 !5 ?4 ❯ make reviewable                                                            
12:41:05 [ OK ] Finished generating deepcopy and crds
sed: 1: "deploy/charts/external- ...": extra characters at the end of d command
sed: 1: "deploy/charts/external- ...": extra characters at the end of d command
sed: 1: "deploy/charts/external- ...": extra characters at the end of d command
sed: 1: "deploy/charts/external- ...": extra characters at the end of d command
make: *** [helm.generate] Error 1
```

-----------

however I still get a couple of other errors for go code especially for these packages `e2e` and `pkg`  - not sure thu if I need to do something locally in my machine to be able to fix/bypass them. please take a look below 


```
12:42:25 [ OK ] Finished generating deepcopy and crds
12:42:25 [ OK ] Finished generating helm chart files
WARN [linters context] gci: `local-prefixes` is deprecated, use `sections` and `prefix(github.com/external-secrets/external-secrets)` instead. 
e2e/suite/import.go:17:1: Expected '\t', Found '\n' at e2e/suite/import.go[line 17,col 1] (gci)

^
pkg/provider/senhasegura/dsm/dsm.go:175:47: httpNoBody: http.NoBody should be preferred to the nil request body (gocritic)
        r, err := http.NewRequest("GET", u.String(), nil)
                                                     ^
pkg/provider/senhasegura/dsm/dsm.go:193:19: ioutilDeprecated: ioutil.ReadAll is deprecated, use io.ReadAll instead (gocritic)
        respData, err := ioutil.ReadAll(resp.Body)
                         ^
pkg/provider/aws/auth/resolver_test.go:48:3: deferInLoop: Possible resource leak, 'defer' is called in the 'for' loop (gocritic)
                defer os.Unsetenv(item.env)
                ^
e2e/suite/aws/secretsmanager/provider.go:27:1: Expected '\t', Found '\n' at e2e/suite/aws/secretsmanager/provider.go[line 27,col 1] (gci)

^
e2e/suite/aws/secretsmanager/secretsmanager.go:18:1: Expected '\t', Found '\n' at e2e/suite/aws/secretsmanager/secretsmanager.go[line 18,col 1] (gci)

^
e2e/suite/aws/secretsmanager/secretsmanager_managed.go:18:1: Expected '\t', Found '\n' at e2e/suite/aws/secretsmanager/secretsmanager_managed.go[line 18,col 1] (gci)

^
e2e/suite/oracle/oracle.go:16:1: Expected '\t', Found '\n' at e2e/suite/oracle/oracle.go[line 16,col 1] (gci)

^
e2e/suite/oracle/provider.go:21:1: Expected '\t', Found '\n' at e2e/suite/oracle/provider.go[line 21,col 1] (gci)

^
e2e/suite/gcp/gcp_managed.go:16:1: Expected '\t', Found '\n' at e2e/suite/gcp/gcp_managed.go[line 16,col 1] (gci)

^
e2e/suite/gcp/provider.go:21:1: Expected '\t', Found '\n' at e2e/suite/gcp/provider.go[line 21,col 1] (gci)

^
pkg/provider/yandex/certificatemanager/certificatemanager_test.go:173:21: stringConcatSimplify: suggestion: certificate1 + "\n" + certificate2 (gocritic)
                strings.TrimSpace(strings.Join([]string{certificate1, certificate2}, "\n")),
                                  ^
pkg/provider/yandex/certificatemanager/certificatemanager_test.go:579:31: stringConcatSimplify: suggestion: certificate1 + "\n" + certificate2 (gocritic)
                        chainProperty:      []byte(strings.Join([]string{certificate1, certificate2}, "\n")),
                                                   ^
pkg/controllers/externalsecret/externalsecret_controller.go:42:1: Expected '\t', Found '\n' at pkg/controllers/externalsecret/externalsecret_controller.go[line 42,col 1] (gci)

^
pkg/controllers/externalsecret/externalsecret_controller_template.go:26:1: Expected '\t', Found '\n' at pkg/controllers/externalsecret/externalsecret_controller_template.go[line 26,col 1] (gci)

^
e2e/suite/template/provider.go:21:1: Expected '\t', Found '\n' at e2e/suite/template/provider.go[line 21,col 1] (gci)

^
e2e/suite/template/template.go:16:1: Expected '\t', Found '\n' at e2e/suite/template/template.go[line 16,col 1] (gci)

^
e2e/suite/akeyless/provider.go:183:15: dynamicFmtString: use errors.New("Require Cloud ID " + err.Error()) or fmt.Errorf("%s", "Require Cloud ID " + err.Error()) instead (gocritic)
                        return "", fmt.Errorf("Require Cloud ID " + err.Error())
                                   ^
e2e/suite/akeyless/akeyless.go:18:1: Expected '\t', Found '\n' at e2e/suite/akeyless/akeyless.go[line 18,col 1] (gci)

^
e2e/suite/akeyless/provider.go:30:1: Expected '\t', Found '\n' at e2e/suite/akeyless/provider.go[line 30,col 1] (gci)

^
e2e/suite/azure/azure_key.go:20:1: Expected '\t', Found '\n' at e2e/suite/azure/azure_key.go[line 20,col 1] (gci)

^
e2e/suite/azure/azure_secret.go:16:1: Expected '\t', Found '\n' at e2e/suite/azure/azure_secret.go[line 16,col 1] (gci)

^
e2e/suite/azure/provider.go:22:1: Expected '\t', Found '\n' at e2e/suite/azure/provider.go[line 22,col 1] (gci)

^
e2e/suite/alibaba/alibaba.go:18:1: Expected '\t', Found '\n' at e2e/suite/alibaba/alibaba.go[line 18,col 1] (gci)

^
e2e/suite/alibaba/provider.go:22:1: Expected '\t', Found '\n' at e2e/suite/alibaba/provider.go[line 22,col 1] (gci)

^
pkg/controllers/secretstore/clustersecretstore_controller.go:29:1: Expected '\t', Found '\n' at pkg/controllers/secretstore/clustersecretstore_controller.go[line 29,col 1] (gci)

^
pkg/controllers/secretstore/secretstore_controller.go:29:1: Expected '\t', Found '\n' at pkg/controllers/secretstore/secretstore_controller.go[line 29,col 1] (gci)

^
e2e/suite/gitlab/gitlab.go:21:1: Expected '\t', Found '\n' at e2e/suite/gitlab/gitlab.go[line 21,col 1] (gci)

^
e2e/suite/gitlab/provider.go:23:1: Expected '\t', Found '\n' at e2e/suite/gitlab/provider.go[line 23,col 1] (gci)

^
pkg/provider/akeyless/akeyless_api.go:55:15: dynamicFmtString: use errors.New("Require Cloud ID " + err.Error()) or fmt.Errorf("%s", "Require Cloud ID " + err.Error()) instead (gocritic)
                        return "", fmt.Errorf("Require Cloud ID " + err.Error())
                                   ^
pkg/provider/akeyless/akeyless_api.go:61:64: response body must be closed (bodyclose)
        authOut, _, err := a.RestAPI.Auth(ctx).Body(*authBody).Execute()
                                                                      ^
pkg/provider/akeyless/akeyless_api.go:103:66: response body must be closed (bodyclose)
        gsvOut, _, err := a.RestAPI.DescribeItem(ctx).Body(body).Execute()
                                                                        ^
pkg/provider/akeyless/akeyless_api.go:127:75: response body must be closed (bodyclose)
        gsvOut, _, err := a.RestAPI.GetRotatedSecretValue(ctx).Body(body).Execute()
                                                                                 ^
pkg/provider/akeyless/akeyless_api.go:172:75: response body must be closed (bodyclose)
        gsvOut, _, err := a.RestAPI.GetDynamicSecretValue(ctx).Body(body).Execute()
                                                                                 ^
pkg/provider/akeyless/akeyless_api.go:202:71: response body must be closed (bodyclose)
        gsvOut, _, err := a.RestAPI.GetSecretValue(ctx).Body(gsvBody).Execute()
                                                                             ^
pkg/provider/senhasegura/auth/iso.go:134:19: ioutilDeprecated: ioutil.ReadAll is deprecated, use io.ReadAll instead (gocritic)
        respData, err := ioutil.ReadAll(resp.Body)
                         ^
e2e/suite/aws/parameterstore/parameterstore.go:18:1: Expected '\t', Found '\n' at e2e/suite/aws/parameterstore/parameterstore.go[line 18,col 1] (gci)

^
e2e/suite/aws/parameterstore/parameterstore_managed.go:18:1: Expected '\t', Found '\n' at e2e/suite/aws/parameterstore/parameterstore_managed.go[line 18,col 1] (gci)

^
e2e/suite/aws/parameterstore/provider.go:26:1: Expected '\t', Found '\n' at e2e/suite/aws/parameterstore/provider.go[line 26,col 1] (gci)

^
e2e/framework/addon/vault.go:33:1: Expected '\t', Found '\n' at e2e/framework/addon/vault.go[line 33,col 1] (gci)

^
e2e/framework/testcase.go:69:4: deferInLoop: Possible resource leak, 'defer' is called in the 'for' loop (gocritic)
                        defer func() {
                        ^
e2e/framework/framework.go:17:1: Expected '\t', Found '\n' at e2e/framework/framework.go[line 17,col 1] (gci)

^
pkg/utils/utils.go:18:1: Expected '\t', Found '\n' at pkg/utils/utils.go[line 18,col 1] (gci)

^
cmd/root.go:27:53: Expected '\t', Found '\n' at cmd/root.go[line 27,col 53] (gci)
        clientgoscheme "k8s.io/client-go/kubernetes/scheme"
                                                           ^
e2e/suite/vault/provider.go:22:1: Expected '\t', Found '\n' at e2e/suite/vault/provider.go[line 22,col 1] (gci)
```

**Comment by msolimans** [comment]
@moolen @gusfcarvalho I added missing `and` for pdb templates in 2 places hopefully this will be the last time to run lint-and-test :) thanks for your time and efforts

**Comment by gusfcarvalho** [comment]
I think your error now might be related either with using go1.18 or another version of golangci-lint :) if you set it to go1.17 and golangci-lint 1.42.1IIRC that should do it

**Comment by msolimans** [comment]
thanks @gusfcarvalho - could you merge since all checks have passed - when can we expect this to go with the next release?

File: deploy/charts/external-secrets/values.yaml [hunk 1]
```diff
@@ -244,6 +255,12 @@ certController:
     # -- Pod priority class name.
   priorityClassName: ""
 
+  # -- Pod disruption budget - for more details see https://kubernetes.io/docs/concepts/workloads/pods/disruptions/
+  podDisruptionBudget:
```
Comment (moolen) [inline deploy/charts/external-secrets/values.yaml:259]: certController doesn't have a way to set the replicaCount. without that it doesn't make sense to add a pdb for it.
Can you add `certController.replicaCount`?

Comment (msolimans) [reply deploy/charts/external-secrets/values.yaml:260]: it makes sense - done

Comment (msolimans) [reply deploy/charts/external-secrets/values.yaml:260]: done

**Comment by msolimans** [comment]
Thanks @moolen for testing that out - I fixed indentation and also added `replicaCount` to `certController`

**Review by moolen — APPROVED** [review]
Looks good, thank you for your contribution!

**Comment by gusfcarvalho** [comment]
@moolen just for us to remember ourselves to put in the release notes this feature is available for kubernetes 1.21 only

**Comment by moolen** [comment]
I can take a look at backporting this to pre-1.21. It should just be a change in the apiversion.

**Comment by gusfcarvalho** [comment]
Not sure if worth it… 1.20 will be rotating out soon

**Pull request merged**